### PR TITLE
Time - Kate P.

### PR DESCRIPTION
**Pull request by Catherina87**
# Task List
Congratulations! You're submitting your assignment!

## Comprehension Questions
Question | Answer
:------------- | :-------------
Describe in your own words what the Model is doing in Rails | Model is similar to ruby class. Model is connected to a table in database. For example if we have a model called Task, the model will be connected with a tasks table. In Rails, Modal (active record) helps to convert tables that live in the database into ruby classes that can be used in ruby code. Tables map to classes, rows map to objects and columns map to object attributes.
Describe in your own words what the Controller is doing in Rails | Controller is a part of the MVC architecture. When Rails app receives an http request, it goes to the routes.rb file, finds the route. The route refers a controller with specific action. Then that action is executed: controller gets the needed data from the Model, an passes the data to the View so that the data can be rendered as HTML.
Describe in your own words what the View is doing in Rails | Receives data from Controller and renders it as HTML. View does not contain any business logic.
Describe an edge-case controller test you wrote | Respond with not found if a user wants to delete a task with invalid id.
What is the purpose of using strong params? (i.e. the params method in the controller) | Strong params provide a safe way to require and permit data that comes from the forms.
How are Rails migrations related to Rails models? | Rails migrations modify the schema, which means that we can add, remove or update columns. And columns in database map to object attributes. So when we create an instance of a Model and we made a migration by adding a new column, we end up having a new attribute.
Describe one area of Rails that are still unclear on | Creating migrations

### CODE REVIEW

File: config/routes.rb [hunk 1]
```diff
@@ -0,0 +1,22 @@
+Rails.application.routes.draw do
+  # For details on the DSL available within this file, see https://guides.rubyonrails.org/routing.html
+  root to: 'tasks#index'
+
```
Comment (beccaelenzil) [inline config/routes.rb:4]: Minor note: style wise it's recommended not to put spaces between your routes.

File: config/routes.rb [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+  get '/tasks/:id', to: 'tasks#show', as: 'task'
+
+  get '/tasks/:id/edit', to: 'tasks#edit', as: 'edit_task'
+
+  patch '/tasks/:id', to: 'tasks#update'
+
+  delete '/tasks/:id', to: 'tasks#destroy', as: 'destroy_task'
+
+  post '/tasks/:id/complete', to: 'tasks#complete', as: 'complete_task'
```
Comment (beccaelenzil) [inline config/routes.rb:20]: Consider how making two separate methods, for instance `mark_complete` and `mark_incomplete` would make this action idempotent.

File: app/views/tasks/new.html.erb [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+    <%= f.label :name %>
+    <%= f.text_field :name %>
+  </div>
+
+  <div>
+    <%= f.label :description %>
+    <%= f.text_field :description %>
+  </div>
+
+  <div>
```
Comment (beccaelenzil) [inline app/views/tasks/new.html.erb:14]: Consider leaving this field off your create form so you have an action to update this attribute.

File: test/controllers/tasks_controller_test.rb [hunk 4]
```diff
@@ -88,33 +88,74 @@
...
-      # Your code here
+      get edit_task_path(task.id)
+
+      must_respond_with :success
     end
     
     it "will respond with redirect when attempting to edit a nonexistant task" do
-      skip
-      # Your code here
+      get edit_task_path(128)
```
Comment (beccaelenzil) [inline test/controllers/tasks_controller_test.rb:97]: consider using a negative number since this will always be invalid. If you have a really large test database, an id of 128 could end up being valid!

File: test/controllers/tasks_controller_test.rb [hunk 5]
```diff
@@ -88,33 +88,74 @@
...
+
+      must_respond_with :redirect
     end
   end
   
   # Uncomment and complete these tests for Wave 3
   describe "update" do
```
Comment (beccaelenzil) [inline test/controllers/tasks_controller_test.rb:104]: Nice working fleshing out the update and destroy tests.

**Comment by beccaelenzil** [comment]
# Task List




## Major Learning Goals/Code Review





| Criteria | yes/no, and optionally any details/lines of code to reference |
| --- | --- |
| At least 6 commits with meaningful commit messages | ✔️
| Routes follow RESTful conventions | ✔️
| Uses named routes (like `_path`) | ✔️
| Creates Models and migrations | ✔️
| Creates styled views | ✔️
| Handles errors like nonexistant tasks | ✔️
| Uses `form_with` to render forms in Rails | ✔️

## Functional Requirements/Manual Testing

| Functional Requirement | yes/no |
| --- | --- |
| Successfully handles index & show | ✔️
| index & show tests pass | ✔️
| Successfully handles: New, Create | ✔️
| New, Create tests pass | ✔️
| Successfully handles: Edit, Update | ✔️
| Successfully handles: Destroy, Task Complete | ✔️

## Overall Feedback

| Overall Feedback | Criteria | yes/no |
| --- | --- | --- |
| Green (Meets/Exceeds Standards) | 5+ in Code Review && 5+ in Functional Requirements | ✔️






## Code Style Bonus Awards



Was the code particularly impressive in code style for any of these reasons (or more...?)

| Quality | Yes? |
| --- | --- |
| Descriptive/Readable | ✅
| Concise | ✅
| Logical/Organized | ✅

**Review by beccaelenzil — COMMENTED** [review]
Great work on this project! It is clear that the learning goals for your first rails project were met. You're off to a great start. Keep up the hard work.
